### ISSUE TITLE
Invariant violation: Reaction with id X did not have 1 author

### ISSUE THREAD

**Issue by jakubgs** [sourcecred/sourcecred#1398]
Full error:
```
IssueComment[MDEyOklzc3VlQ29tbWVudDUyNDg5Njc4MQ==].reactions: unexpected null value
Error: Invariant violation: Reaction with id MDg6UmVhY3Rpb241MDkxMzcxOQ== did not have 1 author
    at RelationalView._addReaction (webpack:///./src/plugins/github/relationalView.js?:61:2803)
    at entry.reactions.expectAllNonNull.map.x (webpack:///./src/plugins/github/relationalView.js?:61:2538)
    at Array.map (<anonymous>)
    at RelationalView._addComment (webpack:///./src/plugins/github/relationalView.js?:61:2526)
    at entry.comments.expectAllNonNull.map.x (webpack:///./src/plugins/github/relationalView.js?:61:721)
    at Array.map (<anonymous>)
    at RelationalView._addPull (webpack:///./src/plugins/github/relationalView.js?:61:709)
    at entry.pulls.expectAllNonNull.map.x (webpack:///./src/plugins/github/relationalView.js?:35:5717)
    at Array.map (<anonymous>)
    at RelationalView._addRepo (webpack:///./src/plugins/github/relationalView.js?:35:5705)
```
Source:
https://github.com/sourcecred/sourcecred/blob/74e7508f11d31b92ad6e3b1cf96580737697a17d/src/plugins/github/relationalView.js#L490-L494

Does this really need to be an error that stops the whole data collection process?
I keep encountering this error and I have to keep adding more and more and more IDs to the `BLACKLISTED_IDS` and it never ends.

Shouldn't this be just a warning and then the comment should be ignored? Does it really makes sense to break the whole process because of it?

**Comment by decentralion** [comment]
You're totally right. There's no reason to break the loading pipeline due to this issue. I think when I wrote that code originally, I didn't know if it would ever come up, so I put the invariant error conservatively. Now that we see that it's creating such a headache, we should definitely change it.

The fix should be quite easy. We can change the type signature of `_addReaction` to return `ReactionRecord | null`, and have it warn and return null if the invariant is violated. 

Then, we can have the calling code just filter out nulls:
https://github.com/sourcecred/sourcecred/blob/74e7508f11d31b92ad6e3b1cf96580737697a17d/src/plugins/github/relationalView.js#L340-L342

Would you be willing to fix this, @jakubgs? If you do, you can edit [this initiative](https://discourse.sourcecred.io/t/make-the-github-plugin-robust/256) to include a link to your pull request, and once I turn on [initiative bounties](https://discourse.sourcecred.io/t/enable-initiative-bounties/257) you'll earn a substantial amount of cred. 

The pre-existing test code does not give a good example for testing this change, because it's all based on snapshots of a known good project, rather than testing behavior on specifically constructed GraphQL responses. 

https://github.com/sourcecred/sourcecred/blob/74e7508f11d31b92ad6e3b1cf96580737697a17d/src/plugins/github/relationalView.test.js#L291-L303

It should be feasible to add a test case by constructing a minimal [Repository](https://github.com/sourcecred/sourcecred/blob/74e7508f11d31b92ad6e3b1cf96580737697a17d/src/plugins/github/graphqlTypes.js#L179) that contains a reaction without an author, then construct a test RelationalView around that repository. I'd be appreciate if you take this approach. However, since this fix has a low probability of regression (and an even lower probability of silent failures), I would also approve and merge a PR that just made the change without updating test code.

**Comment by wchargin** [comment]
This should be fixed by #998. I can reproduce the issue from a fresh
load at time of writing at f86ee92f (prior to the fix), and can load the
repo successfully at eebacff1 (after the fix). Both loading and analysis
complete normally, and the UI shows that the graph correctly includes a
reaction edge from the bot to the comment:

![Screenshot of `dependabot-preview`’s reaction to a comment on the
issue in question][ss]

The relational view still throws an exception if we can’t find the
author of a reaction, but a bot reacting to a comment no longer puts us
into that problematic state, and I’m not aware of any other way to get
into it.

If this recurs with latest SourceCred, please post back here and I’ll
take a look. Thanks!

[ss]: https://user-images.githubusercontent.com/4317806/75619771-009e7380-5b35-11ea-8b8b-2f62b9cb2018.png
